spec/views/components/autocomplete_spec.rb

```diff
@@ -0,0 +1,90 @@
+  end
+
+  it "should not render when only a label is provided" do
+    assert_empty render_component(label: { text: "foo" })
+  end
+
+  it "should render text input with randomised id when name & label provided" do
+    render_component(name: "foo", label: { text: "bar" })
+
+    assert_select ".app-c-autocomplete", count: 1
+    randomised_for = nil
+    assert_select ".app-c-autocomplete label", count: 1, text: "bar" do |label|
+      randomised_for = label[0]['for']
+    end
+    assert_select ".app-c-autocomplete input[name='foo'][id='#{randomised_for}']", count: 1
+  end
+
+  it "should render text input when id, name & label provided" do
+    render_component(name: "foo", label: { text: "bar" }, id: "baz")
+
+    assert_select ".app-c-autocomplete", count: 1
+    assert_select ".app-c-autocomplete label[for='baz']", count: 1, text: "bar"
+    assert_select ".app-c-autocomplete input[name='foo'][id='baz']", count: 1
+  end
+
+  it "renders datalist element associated with text input when options provided" do
+    render_component(
+      id: "basic-autocomplete",
+      name: "foo",
+      label: { text: "Countries" },
+      input: {
+        options: ["United Kingdom", "United States"]
+      }
+    )
+
+    assert_select ".govuk-label", text: "Countries", for: "basic-autocomplete"
+    assert_select ".app-c-autocomplete input[name='foo'][id='basic-autocomplete'][list='basic-autocomplete-list']", count: 1
+    assert_select "datalist#basic-autocomplete-list", count: 1
+    assert_select "datalist#basic-autocomplete-list option[value='United Kingdom']", count: 1
+    assert_select "datalist#basic-autocomplete-list option[value='United States']", count: 1
+  end
+
+  it "renders a default value if provided" do
+    render_component(
+      id: "foo",
+      name: "foo",
+      label: { text: "Countries" },
+      input: {
+        options: ["United Kingdom", "United States"],
+        value: "Spain"
+      }
+    )
+
+    assert_select ".app-c-autocomplete input[value='Spain'][name='foo'][id='foo'][list='foo-list']", count: 1
+    assert_select "datalist#foo-list option[value='United Kingdom']", count: 1
+    assert_select "datalist#foo-list option[value='United States']", count: 1
+    assert_select "datalist#foo-list option[value='Spain']", count: 0
+  end
+
+  it "renders error messages if passed to the component" do
+    render_component(
+      id: "foo",
+      name: "foo",
+      label: { text: "Label" },
+      error_items: [
+        { text: "Something wrong with A" },
+        { text: "Also something wrong with B" }
+      ]
+    )
+
+    assert_select ".app-c-autocomplete .gem-c-error-message", count: 1 do |error|
+      error_message = error[0].text
+      expect(error_message).to include "Something wrong with A"
+      expect(error_message).to include "Also something wrong with B"
+    end
+  end
+
+  def render_component(arguments)
+    render partial: "components/autocomplete", locals: arguments
+  end
+end
```

Comment:
It looks like there are quite a few different ways of calling the component to invoke features (e.g. default values). The convention is to add documentation about how the component can be used to a YAML file to `app/views/components/docs`. Can you add some docs for the auto-complete component.

Content Publisher has an example that cane be used for inspiration: https://github.com/alphagov/content-publisher/blob/master/app/views/components/docs/autocomplete.yml

Reply:
Should that be in the same commit? It's going to be a big commit once it's fixed up to the previous one that adds the view/JS/CSS!

Reply:
Yes please.

Reply:
If the commit gets too big the JS and CSS can be split out.

Reply:
This was done on Friday 👍